### ISSUE TITLE
Crash on malformed ldap syntax in keycloak_ldap_group_mapper.groups_ldap_filter

### ISSUE THREAD

**Issue by lunderhage** [mrparkers/terraform-provider-keycloak#554]
It looks like the plugin is not able to handle errors regarding syntax problems in groups_ldap_filter in keycloak_ldap_group_mapper:

`Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place
 
│ Error: Plugin did not respond
│ 
│   with module.keycloak_dev.keycloak_ldap_group_mapper.first_user_group,
│   on ../../modules/keycloak/main.tf line 396, in resource "keycloak_ldap_group_mapper" "first_user_group":
│  396: resource "keycloak_ldap_group_mapper" "first_user_group" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.
╵
 
Stack trace from the terraform-provider-keycloak_v3.1.1 plugin:
 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x6edc71]
 
goroutine 53 [running]:
github.com/hashicorp/go-version.(*Version).String(0x0, 0xc0003eb890, 0x2)
	github.com/hashicorp/go-version@v1.2.0/version.go:359 +0x51
github.com/hashicorp/go-version.(*Version).Compare(0x0, 0xc000331b30, 0xc0000ec500)
	github.com/hashicorp/go-version@v1.2.0/version.go:118 +0x2f
github.com/hashicorp/go-version.(*Version).GreaterThanOrEqual(...)
	github.com/hashicorp/go-version@v1.2.0/version.go:293
github.com/mrparkers/terraform-provider-keycloak/keycloak.(*KeycloakClient).VersionIsGreaterThanOrEqualTo(0xc00076c660, 0x115178f, 0x6, 0xf96380)
	github.com/mrparkers/terraform-provider-keycloak/keycloak/version.go:20 +0x65
github.com/mrparkers/terraform-provider-keycloak/provider.getLdapGroupMapperFromData(0xc00076c660, 0xc0000f7000, 0xc0001f16c0)
	github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:158 +0xabf
github.com/mrparkers/terraform-provider-keycloak/provider.resourceKeycloakLdapGroupMapperUpdate(0xc0000f7000, 0x114e020, 0xc00076c660, 0x0, 0xffffffffffffffff)
	github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:231 +0x4e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc0005e8580, 0x12fb060, 0xc000317600, 0xc0000f7000, 0x114e020, 0xc00076c660, 0x0, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/helper/schema/resource.go:293 +0x88
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0005e8580, 0x12fb060, 0xc000317600, 0xc00027e310, 0xc0003b6de0, 0x114e020, 0xc00076c660, 0x0, 0x0, 0x0, ...)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/helper/schema/resource.go:395 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc00000cbc0, 0x12fb060, 0xc000317600, 0xc00027e150, 0xc00000cbc0, 0xc00000cbd0, 0x11a6070)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/internal/helper/plugin/grpc_provider.go:952 +0x8cf
github.com/hashicorp/terraform-plugin-sdk/v2/internal/tfplugin5._Provider_ApplyResourceChange_Handler.func1(0x12fb060, 0xc000317600, 0x1101400, 0xc00027e150, 0xc000317600, 0x103d3e0, 0x1, 0xc00009da00)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/internal/tfplugin5/tfplugin5.pb.go:3312 +0x89
github.com/hashicorp/terraform-plugin-sdk/v2/plugin.Serve.func3.1(0x12fb120, 0xc000398390, 0x1101400, 0xc00027e150, 0xc00009d9e0, 0xc00009da00, 0xc000244ba0, 0x5d30c6, 0x10c1a20, 0xc000398390)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/plugin/serve.go:76 +0x87
github.com/hashicorp/terraform-plugin-sdk/v2/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x11173c0, 0xc00000cbc0, 0x12fb120, 0xc000398390, 0xc00076dda0, 0xc00050aba0, 0x12fb120, 0xc000398390, 0xc000140400, 0x83c)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.2-0.20200817173939-b72757e734f6/internal/tfplugin5/tfplugin5.pb.go:3314 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002af880, 0x1305220, 0xc0000b3500, 0xc0000e5500, 0xc0001902d0, 0x19189e0, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.30.0/server.go:1171 +0x522
google.golang.org/grpc.(*Server).handleStream(0xc0002af880, 0x1305220, 0xc0000b3500, 0xc0000e5500, 0x0)
	google.golang.org/grpc@v1.30.0/server.go:1494 +0xcc5
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0000402d0, 0xc0002af880, 0x1305220, 0xc0000b3500, 0xc0000e5500)
	google.golang.org/grpc@v1.30.0/server.go:834 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.30.0/server.go:832 +0x1fd
 
Error: The terraform-provider-keycloak_v3.1.1 plugin crashed!
 
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.`

**Comment by mrparkers** [comment]
Thanks for the bug report.  Would you be able to provide me with some sample HCL to reproduce this issue?

**Comment by lunderhage** [comment]
Om a closer look it seems to have something to do with the update of groups_ldap_filter rather than the syntax (which is might not be executed until a query is passed on to the ldap service).

Here is a sample script:
https://gist.github.com/lunderhage/4aeaa13ec1021027405cb3a8e16f6d0a

Just apply it once, then update groups_ldap_filter and apply again

Keycloak version to reproduce was jboss/keycloak:11.0.2.

**Comment by lunderhage** [comment]
Crash does also occur on deletion.

**Comment by mrparkers** [comment]
Hey @lunderhage, I tested your sample script with the latest version of the provider and v11, v12, and v13 of Keycloak.  I was not able to reproduce this issue, via creating then updating `groups_ldap_filter`, or by deleting everything.

What version of the provider are you currently using?  You can use `terraform version` to check.

**Comment by lunderhage** [comment]
´´´
$ tf version
Terraform v1.0.1
on linux_amd64
+ provider registry.terraform.io/mrparkers/keycloak v3.1.1
´´´

**Comment by lunderhage** [comment]
I am on Ubuntu 20.04 (WSL) if that is of any help:
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04
DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.2 LTS"

**Comment by tomrutsaert** [comment]
And which keycloak version?

**Comment by lunderhage** [comment]
KeyCloak:
Server Version | 11.0.2

**Comment by lunderhage** [comment]
And here is my docker-compose.yml:

version: "3.4"
services:
  keycloak:
    image: jboss/keycloak:11.0.2
    ports:
      - 8070:8080
    # environment:
    #   - KEYCLOAK_USER=admin
    #   - KEYCLOAK_PASSWORD=password

**Comment by lunderhage** [comment]
I managed to reproduce it with v3.2.0 as well:

```
Stack trace from the terraform-provider-keycloak_v3.2.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x74cd31]

goroutine 73 [running]:
github.com/hashicorp/go-version.(*Version).String(0x0, 0xc0003ef1c0, 0x2)
        github.com/hashicorp/go-version@v1.3.0/version.go:371 +0x51
github.com/hashicorp/go-version.(*Version).Compare(0x0, 0xc0000927d0, 0xc0001ea500)
        github.com/hashicorp/go-version@v1.3.0/version.go:118 +0x2f
github.com/hashicorp/go-version.(*Version).GreaterThanOrEqual(...)
        github.com/hashicorp/go-version@v1.3.0/version.go:305
github.com/mrparkers/terraform-provider-keycloak/keycloak.(*KeycloakClient).VersionIsGreaterThanOrEqualTo(0xc000498000, 0x10309ae, 0x6, 0xec0120)
        github.com/mrparkers/terraform-provider-keycloak/keycloak/version.go:21 +0x65
github.com/mrparkers/terraform-provider-keycloak/provider.getLdapGroupMapperFromData(0xc000498000, 0xc000098c80, 0xc000159718)
        github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:158 +0xab8
github.com/mrparkers/terraform-provider-keycloak/provider.resourceKeycloakLdapGroupMapperUpdate(0xc000098c80, 0x102d2a0, 0xc000498000, 0x0, 0xffffffffffffffff)
        github.com/mrparkers/terraform-provider-keycloak/provider/resource_keycloak_ldap_group_mapper.go:231 +0x4e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc000263500, 0x11d31e8, 0xc00039c240, 0xc000098c80, 0x102d2a0, 0xc000498000, 0x0, 0x0, 0x0)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:352 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000263500, 0x11d31e8, 0xc00039c240, 0xc0002a0380, 0xc000786840, 0x102d2a0, 0xc000498000, 0x0, 0x0, 0x0, ...)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:464 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0001a6f60, 0x11d31e8, 0xc00039c240, 0xc000092780, 0xc00039c240, 0xfe5e40, 0xc0002d2600)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/grpc_provider.go:955 +0x8ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc0001cc9c0, 0x11d3290, 0xc00039c240, 0xc0002a00e0, 0xc0001cc9c0, 0xc0002d2630, 0xc00082bba0)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:332 +0xb5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0xfe5e40, 0xc0001cc9c0, 0x11d3290, 0xc0002d2630, 0xc000836240, 0x0, 0x11d3290, 0xc0002d2630, 0xc000244d00, 0x814)
        github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000b36c0, 0x11dba58, 0xc000102d80, 0xc0003a0000, 0xc0003644e0, 0x174eec0, 0x0, 0x0, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0000b36c0, 0x11dba58, 0xc000102d80, 0xc0003a0000, 0x0)
        google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000382270, 0xc0000b36c0, 0x11dba58, 0xc000102d80, 0xc0003a0000)
        google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd

Error: The terraform-provider-keycloak_v3.2.0 plugin crashed!
```

**Comment by lunderhage** [comment]
@mrparkers , Have you been able to reproduce this problem yet?

**Comment by lunderhage** [comment]
With my microscopic knowledge of go I can isolate the problem to github.com/hashicorp/go-version.

If I simply let VersionIsGreaterThanOrEqualTo in version.go (in terraform-provider-keycloak) return true, the plugin succeeds in updating groups_ldap_filter.

So go-version is probably fed with something bad because I can easily reproduce the crash in this sample program:

```
package main

import (
	"bufio"
	"github.com/hashicorp/go-version"
	"os"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	print("Input version 1: ")
	versionString1,_ := reader.ReadString('\n')
	versionString1 = strings.Trim(versionString1, "\n")
	v, _ := version.NewVersion(versionString1)
	println("Version 1: " + v.String())
	print("Input version 2: ")
	versionString2,_ := reader.ReadString('\n')
	versionString2 = strings.Trim(versionString2, "\n")
	v2, _ := version.NewVersion(versionString2)
	println("Version 2: ", v2.String())
	println("Is v greater or equal to v2? ", v.GreaterThanOrEqual(v2))
}
```

And then input something like "11.0.0+" or "11?" to get the same crash on the same line, namely
/go/pkg/mod/github.com/hashicorp/go-version@v1.3.0/version.go:371 +0x51

**Comment by mrparkers** [comment]
That helps a lot!

I was able to narrow down the issue, it was actually because your provider setting had `initial_login` set to `false`.  The login process is what fetches and stores the server version, and the code within the `keycloak_ldap_group_mapper` resource does some version checks when getting and setting data to and from state.  This was happening a few times before the first request was sent to Keycloak, causing a nil pointer to be evaluated.

The fix for this is in #564
